src/loader/chart/utils.ts

```diff
@@ -75,176 +72,148 @@ export const getPrices = async (
-      return sql`date_trunc('hour', ${prices.tickAt}, 'UTC')`;
-    case '1d':
-    case '3d':
-      return sql`date_trunc('day', ${prices.tickAt}, 'UTC')`;
-    case '1w':
-      return sql`date_trunc('week', ${prices.tickAt}, 'UTC')`;
-    case '30d':
-      return sql`date_trunc('month', ${prices.tickAt}, 'UTC')`;
+      return usdDailyPricesTable;
   }
 };
 
-const queryTokenStablecoins = async (
-  baseId: number,
+type PriceData = {
+  tokenId: number;
+  tickAt: Date;
+  value: string;
+  low: string;
+  high: string;
+};
+
+const queryPrices = async (
+  tokenId: number,
   quoteId: number,
-  stablecoinId: number,
   startTimestamp: Date,
   endTimestamp: Date,
   timeframe: Timeframe,
-) =>
-  db
+): Promise<PriceData[]> => {
+  const table = tableByTimeframe(timeframe);
+  return db
     .select({
-      baseId: prices.baseId,
-      quoteId: prices.quoteId,
-      date: timeframeToSql(timeframe).mapWith(String).as('date'),
-      value: prices.value,
+      tokenId: table.tokenId,
+      tickAt: table.tickAt,
+      value: table.value,
+      low: table.low,
+      high: table.high,
     })
-    .from(prices)
+    .from(table)
     .where(
       and(
-        inArray(prices.baseId, [baseId, quoteId]),
-        eq(prices.quoteId, stablecoinId),
-        between(prices.tickAt, startTimestamp, endTimestamp),
+        or(eq(table.tokenId, tokenId), eq(table.tokenId, quoteId)),
+        between(table.tickAt, startTimestamp, endTimestamp),
       ),
-    )
-    .groupBy(prices.baseId, prices.quoteId, sql`date`, prices.value);
+    );
+};
 
-const queryTokenStartPrice = async (baseId: number, quoteId: number, startTimestamp: Date) =>
-  db.query.prices.findFirst({
-    columns: {
-      baseId: true,
-      quoteId: true,
-      value: true,
-    },
-    where: and(eq(prices.baseId, baseId), eq(prices.quoteId, quoteId), lte(prices.tickAt, startTimestamp)),
-    orderBy: desc(prices.tickAt),
-  });
+const getTokenStartPrice = async (
+  tokenId: number,
+  beforeTimestamp: Date,
+  timeframe: Timeframe,
+): Promise<PriceData[]> => {
+  const table = tableByTimeframe(timeframe);
+  return db
+    .select({ tokenId: table.tokenId, tickAt: table.tickAt, value: table.value, low: table.low, high: table.high })
+    .from(table)
+    .where(and(eq(table.tokenId, tokenId), lte(table.tickAt, beforeTimestamp)))
+    .orderBy(desc(table.tickAt))
+    .limit(1);
+};
 
-const getTokenIds = async (chainId: number, baseTokenAddress: string, quoteTokenAddress: string) => {
-  const stablecoinAddress = networks.getByChainId(chainId)?.stablecoinAddress.toLowerCase();
+const validateStartPrice = async (items: PriceData[], tokenId: number, startTimestamp: Date, timeframe: Timeframe) => {
+  if (items.find((item) => item.tokenId === tokenId && dayjs(item.tickAt).isSame(startTimestamp)) === undefined) {
+    const startData = await getTokenStartPrice(tokenId, startTimestamp, timeframe);
+    if (startData.length) {
+      items.unshift(...startData);
```

Comment:
You are modifying the existing array. Don't you want rather to return the new array?

Reply:
Yes, that's intentional - if there is missing prices for token, i'm pushing them to the front of the existing array there.
This function can be reused multiple times for the same array (for different tokens), so looked better than having to re-reset original variable in parent function with new array